### PR TITLE
#5519 [a11y] Button announces multiple times when status changes to isPending

### LINKED ISSUE #5519: Button announces multiple times when status changes to isPending
### Provide a general summary of the issue here

Following on #5245, Button announces more than once when the status changes to `isPending={true}` using Firefox with NVDA or JAWS.

### 🤔 Expected Behavior?

Button should not announce more than once when the status changes to `isPending={true}`.

### 😯 Current Behavior

Button announces more than once when the status changes to `isPending={true}` with NVDA and JAWS.

### 💁 Possible Solution

_No response_

### 🔦 Context

_No response_

### 🖥️ Steps to Reproduce

1. Open https://reactspectrum.blob.core.windows.net/reactspectrum/88c4d560d552d6d911a71396759f9088b1df2199/storybook/index.html?path=/story/button--pending-spinner&providerSwitcher-express=false example in Firefox on Windows using JAWS or NVDA. 
2. Press any of the isPending button examples. 
3. When the button enters the pending state, the screen reader announces the button name and "pending" more than once.

### Version

https://github.com/adobe/react-spectrum/commit/9d8c74a344c9b7539f00fa58d0a02b062b6e830e

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Microsoft Edge

### If other, please specify.

_No response_

### What operating system are you using?

Windows, macOS

### 🧢 Your Company/Team

Adobe/Accessibility

### 🕷 Tracking Issue

_No response_

### PR DESCRIPTION
**Pull request by majornista**
Closes #5519

## ✅ Pull Request Checklist:

- [x] Included link to corresponding React Spectrum GitHub Issue #5519.
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. Open https://reactspectrum.blob.core.windows.net/reactspectrum/399b9f9af6f93a7b8f93327aa70aa7d31d9a4fd5/storybook/iframe.html?providerSwitcher-express=false&scrolling=false&providerSwitcher-locale=&providerSwitcher-theme=&providerSwitcher-scale=&args=isPending:true&id=button--pending-spinner&viewMode=story in Firefox on Windows using NVDA or JAWS.
2. Verify that clicking each of the buttons in the example announces using the buttons accessibility name and "pending" once when the `isProgressVisible` state enters.
3. Verify that the Controlled example is named "Controlled pending" when `isPending` is set to `true` in the Actions panel 
4. Repeat using Chrome and Microsoft Edge on Windows, and using Chrome and Safari with VoiceOver on macOS.
5. Note that using Safari with VoiceOver, you can use Control + Option + Space to trigger a virtual click with the screen reader. This seems to make VoiceOver announce the "pending" state more consistently than pressing Enter or Space on their own.

## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

**Comment by majornista** [comment]
> magic
> 
> it's working well for me across Chrome/Safari MacOS and iOS

Trial and error. I'll add some comments to explain the changes.

File: packages/@adobe/spectrum-css-temp/components/button/index.css [hunk 1]
```diff
@@ -167,7 +167,7 @@ governing permissions and limitations under the License.
 
     .spectrum-Button-label,
     .spectrum-Icon {
-      visibility: hidden;
+      opacity: 0;
```
Comment (majornista) [inline packages/@adobe/spectrum-css-temp/components/button/index.css:170]: `visibility: hidden` hides the element from assistive technology so when the IDREF for the label or icon is used in `aria-labelledby` the accessibility name for the icon or label will be omitted from the accessibility name for the element referencing it. Using `opacity: 0` ensures that the elements will still render to the DOM and be available for reference using `aria-labelledby`.

File: packages/@react-aria/utils/src/index.ts [hunk 2]
```diff
@@ -30,7 +30,7 @@ export {useSyncRef} from './useSyncRef';
 export {getScrollParent, isScrollable} from './getScrollParent';
 export {useViewportSize} from './useViewportSize';
 export {useDescription} from './useDescription';
-export {isMac, isIPhone, isIPad, isIOS, isAppleDevice, isWebKit, isChrome, isAndroid} from './platform';
+export {isMac, isIPhone, isIPad, isIOS, isAppleDevice, isWebKit, isChrome, isAndroid, isFirefox} from './platform';
```
Comment (majornista) [inline packages/@react-aria/utils/src/index.ts:33]: `isFirefox` was omitted from the public API.

File: packages/@react-spectrum/button/src/Button.tsx [hunk 3]
```diff
@@ -78,10 +78,9 @@ function Button<T extends ElementType = 'button'>(props: SpectrumButtonProps<T>,
   let [isProgressVisible, setIsProgressVisible] = useState(false);
   let backupButtonId = useId();
   let buttonId = buttonProps.id || backupButtonId;
-  let spinnerId = useId();
-  let textId = useId();
   let iconId = useId();
-  let auxLabelId = useId();
+  let textId = useId();
+  let spinnerId = useId();
```
Comment (majornista) [inline packages/@react-spectrum/button/src/Button.tsx:83]: I'm weird, and felt compelled to generate ids in the order they appear in the DOM, which helps make debugging in the Web Inspector a little easier.

File: packages/@react-spectrum/button/src/Button.tsx [hunk 4]
```diff
@@ -119,9 +125,8 @@ function Button<T extends ElementType = 'button'>(props: SpectrumButtonProps<T>,
         data-style={style}
         data-static-color={staticColor || undefined}
         aria-disabled={isPending ? 'true' : undefined}
-        aria-label={isPending ? undefined : buttonProps['aria-label']}
-        aria-labelledby={isPending ? undefined : buttonProps['aria-labelledby']}
-        aria-live={isPending && isFocused ? 'polite' : undefined}
+        aria-label={isPending ? isPendingAriaLiveLabel : buttonProps['aria-label']}
+        aria-labelledby={isPending ? isPendingAriaLiveLabelledby : buttonProps['aria-labelledby']}
```
Comment (majornista) [inline packages/@react-spectrum/button/src/Button.tsx:129]: Using `aria-label={isPendingAriaLiveLabel}` and `aria-labelledby={isPendingAriaLiveLabelledby}` on the button element ensures that "pending" is included in the button's accessibility name when `isPending` is `true`.

File: packages/@react-spectrum/button/src/Button.tsx [hunk 5]
```diff
@@ -153,23 +158,26 @@ function Button<T extends ElementType = 'button'>(props: SpectrumButtonProps<T>,
...
+                {isProgressVisible &&
+                  <div role="img" aria-labelledby={isPendingAriaLiveLabelledby} />
+                }
+              </div>
+              {/* Adding the element here with the same labels as the button itself causes aria-live to pick up the change in Safari.
+              Safari with VO unfortunately doesn't announce changes to *all* of its labels specifically for button
+              https://a11ysupport.io/tests/tech__html__button-name-change#assertion-aria-aria-label_attribute-convey_name_change-html-button_element-vo_macos-safari
+              The aria-live may cause extra announcements in other browsers. */}
+              <div id={spinnerId} role="img" aria-label={isPendingAriaLiveLabel} />
+            </>
```
Comment (majornista) [inline packages/@react-spectrum/button/src/Button.tsx:180]: The `aria-live` region is added when `isPending` is `true`, but we don't add the child element until `isProgressVisible` is also `true`, because the delay between adding `aria-live` element and its child helps ensure that the screen reader picks up on the live region before announcing the addition.

The `spinnerId` element is outside the `aria-live` region, so that it is not considered another addition to the live region requiring announcement.  

Since we are adding `aria-label={isPendingAriaLiveLabel}` and `aria-labelledby={isPendingAriaLiveLabelledby}` to the parent button when `isPending` is `true`, we don't need to set `aria-hidden="true"` on this element.

Comment (snowystinger) [reply packages/@react-spectrum/button/src/Button.tsx:180]: i fought with this a few times trying to get it to work, some browsers seemed fine if live plus its contents were added at the same time, others not so much. great idea moving the live region down, I hadn't thought to do that

File: packages/@adobe/spectrum-css-temp/components/button/index.css (hunk 1 above)
Comment (majornista) [reply packages/@adobe/spectrum-css-temp/components/button/index.css:170]: It kind of caught me by surprise too.

**Review by yihuiliao — APPROVED** [review]
it works so well now!

i also tested it on VO Chrome/Safari, NVDA Chrome/Firefox/Edge

**Review by LFDanLu — APPROVED** [review]
Verified the behavior on Android Talkback, VO, and NVDA on various browsers, much improved!
